### ISSUE TITLE
Define the JSON WireService persistence form

### ISSUE THREAD

**Issue by MMaiero** [eclipse/kura#1808]
A new JSON persistence representation needs to be defined in order to contain both the nodes and arcs of the WireGraph. The JSON can also contain graphical details used to render the graph (i.e. component position in the canvas, rotation, number of ports, etc.)

**Comment by nicolatimeus** [comment]
First proposal:
```json
{
  "components": [{
      "pid": "foo",
      "inputPortCount": 0,
      "outputPortCount": 5,
      "renderingProperties": {
        "position": {
          "x": 10,
          "y": 100
        },
        "inputPortNames": {
          "0": "resetPort"
        },
        "outputPortNames": {
          "3": "then",
          "4": "else"
        },
        "rotation": 0,
        "color": "red",
        "size": {
          "width": 100,
          "height": 50
        }
      }
    },
    {
      "pid": "bar",
      "inputPortCount": 6,
      "outputPortCount": 1,
      "renderingProperties": {
        "position": {
          "x": 150,
          "y": 100
        },
        "rotation": 0,
        "color": "red",
        "size": {
          "width": 100,
          "height": 50
        }
      }
    }
  ],
  "wires": [{
    "emitter": "foo",
    "receiver": "bar",
    "emitterPort": 0,
    "receiverPort": 5
  }],
}
```
The above json will be persisted in the WireServiceImpl configuration inside a single property. Since it also contains the description of the wires, it will be possible to remove the current wire related properties from WireServiceImpl.

* Wire component properties:
  * `pid`: The `kura.service.pid` of the component (mandatory)
  * `inputPortCount/outputPortCount`: The number of input/output ports of the component, if the corresponding value is greater than zero, the component is a `WireReceiver/WireEmitter`. (mandatory)
  * `renderingProperties`: some representation-related properties (mandatory)
    * `position`: the position of the component inside the canvas (mandatory)
    * `inputPortNames/outputPortNames`: the labels of the input/output ports, if not provided the ports will have no label, it is also possible to specify labels only for a partial set of ports (optional)
    * all other parameters are included only as an example of other possible rendering properties and will be ignored by the initial implementation

 * Wire properties:
   * `emitter/receiver`: the `kura.service.pid` of the emitter/receiver component (mandatory)
   * `emitterPort/receiverPort`: the port on the emitter/receiver component to which this wire is attached, if these parameters are not set, it will be assumed that the wire is connected to port 0. (optional)

**Comment by MMaiero** [comment]
@aabramo Do you have any comments on this?

**Comment by aabramo** [comment]
As far I understand, this would be an intermediate representation from the Kura "internal" one and the rendering package, JointJs or whatever else.
The strategy is very suggestive. The only concern I have is on the possibility/risk of being exhaustive/partial including all features required to encompass the whole set of graphical variations we would like to be implemented (in future; just to mention, hierarchical blocks).
Provided that this representation can be extended (modified?), I don't see flaws discouraging its adoption.

**Comment by MMaiero** [comment]
Yes, the idea is to create something that can be easily extended. 
The current idea is to use this proposed JSON format only for persistence.

The aim now is exactly to try to think at possible extensions of the Wires and evaluate if them can be covered by this proposal (or an extended version) without the need to rewrite everything.

**Comment by MMaiero** [comment]
Merged PR #1819
